modules/core/src/core/lifecycle/create-props.js

```diff
@@ -106,25 +112,92 @@ function buildPropDefs(layerClass) {
   return {propTypes, defaultProps};
 }
 
-function buildDefaultProps(props, parentProps, propTypes, layerClass) {
+// Builds a pre-merged default props object that layer props can inherit from
+function buildDefaultPropsObject(props, parentProps, propTypes, layerClass) {
   const defaultProps = Object.create(null);
 
   Object.assign(defaultProps, parentProps, props);
 
-  const descriptors = {};
-
+  // Avoid freezing `id` prop
   const id = getLayerName(layerClass);
   delete props.id;
 
-  Object.assign(descriptors, {
+  // Add getters/setters for async prop properties
+  Object.defineProperties(defaultProps, {
     id: {
       configurable: false,
       writable: true,
       value: id
     }
   });
 
-  Object.defineProperties(defaultProps, descriptors);
+  // Add getters/setters for async prop properties
+  addAsyncPropDescriptors(defaultProps, propTypes);
 
   return defaultProps;
 }
+
+// Create descriptors for overridable props
+function addAsyncPropDescriptors(defaultProps, propTypes) {
+  const defaultValues = {};
+
+  const descriptors = {
+    // Default "resolved" values for async props, returned if value not yet resolved/set.
+    _asyncPropDefaultValues: {
+      enumerable: false,
+      value: defaultValues
+    },
+    // TODO - Shadowed object, just to allow indexing
+    _asyncPropUrls: {
+      enumerable: false,
+      value: {}
+    }
+  };
+
+  // Move async props into shadow values
+  for (const propType of Object.values(propTypes)) {
```

Comment:
What is the difference between this and for...in?

Reply:
Some minor advantages: It returns the value rather than the key, so it removes a line of code needed to dereference the object, and also uses the arguably more common `for-of` loop construct. 

But there was no particular reason for picking this syntax, and it is true that we don't use it much in the code base. I will switch to `for-in` for consistency.